### ISSUE TITLE
LPIPS Loss producing negative values

### ISSUE THREAD

**Issue by GuillaumeRochette** [richzhang/PerceptualSimilarity#72]
Hi,

While running the LPIPS loss based on AlexNet, I obtained a negative value,
```python
a = LPIPS(net="alex", verbose=False)
x = torch.rand(4, 3, 256, 256)
y = torch.rand(4, 3, 256, 256)
z = a(x, y, normalize=True)
print(z)
```
While looking at the values contained in `res` (defined in the `forward()`), I have noticed that the implementation does not match the `Eq. 1` from the paper.

Here's Eq. 1:
![image](https://user-images.githubusercontent.com/11836548/125159455-5e05f400-e16f-11eb-916d-214eec17cbbd.png)

While this is what is implemented,
![image](https://user-images.githubusercontent.com/11836548/125159495-a02f3580-e16f-11eb-82cd-61d6adad0f6f.png)

The square operation `** 2` at line 94 should be removed and instead applied on the `self.lins[kk].model(diffs[kk])` (at lines 98 and 100), and on `diff[kk]` (at lines 103 and 105).

Thanks in advance,

Guillaume

**Comment by markdjwilliams** [comment]
Is there a good workaround for this?

**Comment by markdjwilliams** [comment]
Thank you, this makes perfect sense.

**Comment by tomsawyer0224** [comment]
@richzhang, thank you because of this repo!
> If the code is installed and the weights are loaded properly (and weren't changed by accidentally fine-tuning them, for example), it is **not possible to get negative values**.
> 
> Check the weights at all non-negative, by doing the following
> 
> ```
> for ll in range(5):
>     print(loss_fn_vgg.lins[ll].model[1].weight.flatten())
> ```
as you mentioned that: the weight of all lins layers are non-negative, I printed out and this is true but I don't understand why. I read many times and can't found the answer. Could you please explain why the weight of lins layer are non-negative? Are they also "pre-trained"?
Thank again,

**Comment by PENGUINLIONG** [comment]
For those trapped in this same problem, if you load LPIPS with

```python
lpips = lpips.LPIPS(net='vgg')
```

**Remember to refute gradients!** Do the following instead!

```python
lpips = lpips.LPIPS(net='vgg').requires_grad_(False)
```

**Comment by sravan953** [comment]
> For those trapped in this same problem, if you load LPIPS with
> 
> lpips = lpips.LPIPS(net='vgg')
> **Remember to refute gradients!** Do the following instead!
> 
> lpips = lpips.LPIPS(net='vgg').requires_grad_(False)

I am only encountering negative values in DDP, and your fix works. Any ideas as to what is happening? Especially since setting `requires_grad(False)` is taken care of in code:
https://github.com/richzhang/PerceptualSimilarity/blob/082bb24f84c091ea94de2867d34c4544f68e0963/lpips/pretrained_networks.py#L33

**Comment by Grant2Grant** [comment]
> > For those trapped in this same problem, if you load LPIPS with
> > lpips = lpips.LPIPS(net='vgg')
> > **Remember to refute gradients!** Do the following instead!
> > lpips = lpips.LPIPS(net='vgg').requires_grad_(False)
> 
> I am only encountering negative values in DDP, and your fix works. Any ideas as to what is happening? Especially since setting `requires_grad(False)` is taken care of in code:
> 
> [PerceptualSimilarity/lpips/pretrained_networks.py](https://github.com/richzhang/PerceptualSimilarity/blob/082bb24f84c091ea94de2867d34c4544f68e0963/lpips/pretrained_networks.py#L33)
> 
> Line 33 in [082bb24](/richzhang/PerceptualSimilarity/commit/082bb24f84c091ea94de2867d34c4544f68e0963)
> 
>  param.requires_grad = False

With  `lpips = lpips.LPIPS(net='vgg')`
we got the following.

```
for name,p in loss_fn.named_parameters():
...     print(name,p.requires_grad)
...     
net.slice1.0.weight False
net.slice1.0.bias False
net.slice1.2.weight False
net.slice1.2.bias False
net.slice2.5.weight False
net.slice2.5.bias False
net.slice2.7.weight False
net.slice2.7.bias False
net.slice3.10.weight False
net.slice3.10.bias False
net.slice3.12.weight False
net.slice3.12.bias False
net.slice3.14.weight False
net.slice3.14.bias False
net.slice4.17.weight False
net.slice4.17.bias False
net.slice4.19.weight False
net.slice4.19.bias False
net.slice4.21.weight False
net.slice4.21.bias False
net.slice5.24.weight False
net.slice5.24.bias False
net.slice5.26.weight False
net.slice5.26.bias False
net.slice5.28.weight False
net.slice5.28.bias False
lin0.model.1.weight True
lin1.model.1.weight True
lin2.model.1.weight True
lin3.model.1.weight True
lin4.model.1.weight True
```
Although I'm not sure why this is related to negative value.

**Comment by sravan953** [comment]
> > > For those trapped in this same problem, if you load LPIPS with
> > > lpips = lpips.LPIPS(net='vgg')
> > > **Remember to refute gradients!** Do the following instead!
> > > lpips = lpips.LPIPS(net='vgg').requires_grad_(False)
> > 
> > 
> > I am only encountering negative values in DDP, and your fix works. Any ideas as to what is happening? Especially since setting `requires_grad(False)` is taken care of in code:
> > [PerceptualSimilarity/lpips/pretrained_networks.py](https://github.com/richzhang/PerceptualSimilarity/blob/082bb24f84c091ea94de2867d34c4544f68e0963/lpips/pretrained_networks.py#L33)
> > Line 33 in [082bb24](/richzhang/PerceptualSimilarity/commit/082bb24f84c091ea94de2867d34c4544f68e0963)
> > param.requires_grad = False
> 
> With `lpips = lpips.LPIPS(net='vgg')` we got the following.
> 
> ```
> for name,p in loss_fn.named_parameters():
> ...     print(name,p.requires_grad)
> ...     
> net.slice1.0.weight False
> net.slice1.0.bias False
> net.slice1.2.weight False
> net.slice1.2.bias False
> net.slice2.5.weight False
> net.slice2.5.bias False
> net.slice2.7.weight False
> net.slice2.7.bias False
> net.slice3.10.weight False
> net.slice3.10.bias False
> net.slice3.12.weight False
> net.slice3.12.bias False
> net.slice3.14.weight False
> net.slice3.14.bias False
> net.slice4.17.weight False
> net.slice4.17.bias False
> net.slice4.19.weight False
> net.slice4.19.bias False
> net.slice4.21.weight False
> net.slice4.21.bias False
> net.slice5.24.weight False
> net.slice5.24.bias False
> net.slice5.26.weight False
> net.slice5.26.bias False
> net.slice5.28.weight False
> net.slice5.28.bias False
> lin0.model.1.weight True
> lin1.model.1.weight True
> lin2.model.1.weight True
> lin3.model.1.weight True
> lin4.model.1.weight True
> ```
> 
> Although I'm not sure why this is related to negative value.

Was that the output **without** `requires_grad_(False)`? If yes, I see the final classifier layers are set to be trainable which might have led to the negative values.

**Comment by jzt-tohsaka60** [comment]
> > > For those trapped in this same problem, if you load LPIPS with
> > > lpips = lpips.LPIPS(net='vgg')
> > > **Remember to refute gradients!** Do the following instead!
> > > lpips = lpips.LPIPS(net='vgg').requires_grad_(False)
> > 
> > 
> > I am only encountering negative values in DDP, and your fix works. Any ideas as to what is happening? Especially since setting `requires_grad(False)` is taken care of in code:
> > [PerceptualSimilarity/lpips/pretrained_networks.py](https://github.com/richzhang/PerceptualSimilarity/blob/082bb24f84c091ea94de2867d34c4544f68e0963/lpips/pretrained_networks.py#L33)
> > Line 33 in [082bb24](/richzhang/PerceptualSimilarity/commit/082bb24f84c091ea94de2867d34c4544f68e0963)
> > param.requires_grad = False
> 
> With `lpips = lpips.LPIPS(net='vgg')` we got the following.
> 
> ```
> for name,p in loss_fn.named_parameters():
> ...     print(name,p.requires_grad)
> ...     
> net.slice1.0.weight False
> net.slice1.0.bias False
> net.slice1.2.weight False
> net.slice1.2.bias False
> net.slice2.5.weight False
> net.slice2.5.bias False
> net.slice2.7.weight False
> net.slice2.7.bias False
> net.slice3.10.weight False
> net.slice3.10.bias False
> net.slice3.12.weight False
> net.slice3.12.bias False
> net.slice3.14.weight False
> net.slice3.14.bias False
> net.slice4.17.weight False
> net.slice4.17.bias False
> net.slice4.19.weight False
> net.slice4.19.bias False
> net.slice4.21.weight False
> net.slice4.21.bias False
> net.slice5.24.weight False
> net.slice5.24.bias False
> net.slice5.26.weight False
> net.slice5.26.bias False
> net.slice5.28.weight False
> net.slice5.28.bias False
> lin0.model.1.weight True
> lin1.model.1.weight True
> lin2.model.1.weight True
> lin3.model.1.weight True
> lin4.model.1.weight True
> ```
> 
> Although I'm not sure why this is related to negative value.

I think it would be better if you create a help function like:
```
   @staticmethod
    def freeze(model):
        model.eval()
        for p in model.parameters():
            p.requires_grad = False`
```
This would freeze all layers including the linear layers
